javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -153,12 +169,39 @@ private Vector(int indexShift, HashArrayMappedTrie<Integer, T> trie) {
         if (elements instanceof Vector) {
             return (Vector<T>) elements;
         } else {
-            HashArrayMappedTrie<Integer, T> trie = HashArrayMappedTrie.empty();
-            for (T element : elements) {
-                trie = trie.put(trie.size(), element);
-            }
-            return wrap(trie);
+            final Seq<? extends T> seq = seq(elements);
+            return seq.isEmpty() ? empty()
+                                 : ofAll(seq, seq.size());
+        }
+    }
+
+    public static <T> Vector<T> ofAll(Collection<? extends T> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        return elements.isEmpty() ? empty()
+                                  : ofAll(elements, elements.size());
+    }
+
+    public static <T> Vector<T> ofAll(Iterable<? extends T> iterable, int size) {
```

Comment:
This way of creating everything from bottom up and proved to be `4-10`x faster than both `clojure` and `scala`